018-custom-id-support.md

```diff
@@ -0,0 +1,128 @@
+
+## Summary
+
+[summary]: #summary
+
+<!-- Brief explanation of the proposed contribution. Write your answer below. -->
+We are planning to support custom IDs to our resources like collections, documents, function tags and other resources. These will allow developers to provide pretty and understandable names to resources such as collections where IDs are required while making a query request.
+
+## Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+
+**What problem are you trying to solve?**
+
+Developers want to name their resources with their own custom ID that they can easily query with. This will also help make query URL understandable and pretty.
+
+**What is the context or background in which this problem exists?**
+
+At the moment the IDs for all the resources are auto generated by Appwrite. Developers are forced to use the generated IDs in their applications to make queries. The generated ids are random and not easily understandable.
+
+**Once the proposal is implemented, how will the system change?**
+
+Every resource creation API will have an optional **id** attribute. Whenever **id** is not provided Appwrite will still generate the IDs as it is already doing. IDs will not be modifiable once the resource is created.
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+<!--
+This is the technical portion of the RFC. Explain the design in sufficient detail keeping in mind the following:
+
+- Its interaction with other parts of the system is clear
+- It is reasonably clear how the contribution would be implemented
+- Dependencies on libraries, tools, projects or work that isn't yet complete
+- New API routes that need to be created or modifications to the existing routes (if needed)
+- Any breaking changes and ways in which we can ensure backward compatibility.
+- Use Cases
+- Goals
+- Deliverables
+- Changes to documentation
+- Ways to scale the solution
+
+Ensure that you include examples, code-snippets etc. to allow the community to understand the proposed solution. **It would be best if the examples use naming conventions that you intend to use during the actual implementation so that changes can be suggested early on during the development.**
+
+Write your answer below.
+
+-->
+Resources that can have Custom IDs
+- Projects
+- Collections
+- Documents
+- Buckets
+- Files
+- Users
+- Function tags
+- Tasks
+- Webhooks
+
+The create resource endpoints for above resources will have a first required **id** parameter where users can send their custom id following the guidelines for validation. In every endpont, if the **id** parameter is received empty (required for backward compatibility with existing SDKs) or pre defined string, Appwrite will still generate the ids as it is already doing.
+
+We can use the existing key validator to validate the custom id as well.
+
+### Some guidelines for custom ids
+
+- Can be at most 32 characters
```

Comment:
We maybe should allow 36 characters to support GUID's 🤔

Reply:
@kodumbeats you feedback please 🙏🏻

Reply:
According to my research, we should be safe so long as we stay under 64 characters, pretty standard across SQL backends:
- https://mariadb.com/kb/en/identifier-names/#maximum-length
- https://dev.mysql.com/doc/mysql-reslimits-excerpt/5.6/en/identifier-length.html
- https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS
- https://docs.mongodb.com/manual/reference/limits/#naming-restrictions

Reply:
@TorstenDittmann @kodumbeats updated to make it 36 chars